### PR TITLE
Validate template values on calls to the scaffolder API

### PR DESCRIPTION
**Pull request by iain-b**
Calls to the scaffolder API should use the schema from the template
entity to validate input values.

You can run this using the following:
```node
const axios = require("axios");
(async () => {
  try {
    const response = await axios.post(
      "http://localhost:7000/scaffolder/v1/jobs/",
      {
        template: {
          apiVersion: "backstage.io/v1alpha1",
          kind: "Template",
          metadata: {
            description: "Create a new CRA website project",
            name: "create-react-app-template",
            tags: ["experimental", "react", "cra"],
            title: "Create React App Template",
          },
          spec: {
            owner: "web@example.com",
            path: ".",
            schema: {
              properties: {
                component_id: {
                  description: "Unique name of the component",
                  title: "Name",
                  type: "string",
                },
                description: {
                  description: "Description of the component",
                  title: "Description",
                  type: "string",
                },
                use_typescript: {
                  default: true,
                  description: "Include typescript",
                  title: "Use Typescript",
                  type: "boolean",
                },
              },
              required: ["component_id", "use_typescript"],
            },
            templater: "cra",
            type: "website",
          },
        },
        values: {},
      }
    );
    console.log(response.data);
  } catch (e) {
      console.log(e.response.data);
  }
})();
```
output looks like (axios has parsed it content type is json):
```node
{
  errors: [
    {
      property: 'instance',
      message: 'requires property "component_id"',
      schema: [Object],
      instance: {},
      name: 'required',
      argument: 'component_id',
      stack: 'instance requires property "component_id"'
    },
    {
      property: 'instance',
      message: 'requires property "use_typescript"',
      schema: [Object],
      instance: {},
      name: 'required',
      argument: 'use_typescript',
      stack: 'instance requires property "use_typescript"'
    }
  ]
}
```

#### :heavy_check_mark: Checklist

- [x] All tests are passing `yarn test`
- [x] Screenshots attached (for UI changes)
- [ ] Relevant documentation updated
- [x] Prettier run on changed files
- [x] Tests added for new functionality
- [ ] Regression tests added for bug fixes

### CODE REVIEW

File: plugins/scaffolder-backend/package.json [hunk 1]
```diff
@@ -36,6 +36,7 @@
     "git-url-parse": "^11.1.2",
     "globby": "^11.0.0",
     "helmet": "^4.0.0",
+    "jsonschema": "^1.2.6",
```
Comment (iain-b) [inline plugins/scaffolder-backend/package.json:39]: json-schema is used in catalog model but it's not well maintained - the implementation anway. The test case I wrote passed validation (with no input) as it seems it doesn't treat the `required` field in the correct way. I've gone with this one but there doesn't really seem to be a standout that's compatibile with the json-schema standard.

File: plugins/scaffolder-backend/src/service/router.ts [hunk 2]
```diff
@@ -84,6 +85,12 @@ export async function createRouter(
       const values: RequiredTemplateValues & Record<string, JsonValue> =
         req.body.values;
 
+      const validationResult: ValidatorResult = validate(values, template.spec.schema);
+      if (!validationResult.valid) {
+        res.status(400).json({ errors: validationResult.errors });
```
Comment (iain-b) [inline plugins/scaffolder-backend/src/service/router.ts:90]: We could just return validationResult.toString() which creates a human readable list but I've chosen this as I think something more machine readable makes sense here given it's most likely this will be hit during API usage.

File: plugins/scaffolder-backend/src/service/router.test.ts [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+            },
+          },
+          required: ['component_id', 'use_typescript'],
+        },
+        templater: 'cra',
+        type: 'website',
+      },
+    };
+
+    it('rejects template values which do not match the template schema definition', async () => {
```
Comment (iain-b) [inline plugins/scaffolder-backend/src/service/router.test.ts:85]: I've only tested the code I've added as the happy path is not trivial and IMO would require refactoring to write a clean test (and given I'm not familiar with the code base I'm not inclined to)

**Comment by dtuite** [comment]
@iain-b this doesn't actually display an error to the user right? It just validates on the server and returns errors.

Just want to make sure I'm understanding it correctly.

**Review by dtuite — APPROVED** [review]

**Comment by iain-b** [comment]
> @iain-b this doesn't actually display an error to the user right? It just validates on the server and returns errors.
> 
> Just want to make sure I'm understanding it correctly.

TL;DR Yes, this isn't intended to be displayed to the user. 

AFAIK this feature is intended to cater for direct API usage where currently there's no validation. On the frontend the same schema we use here to validate is used to build the form, so barring a bug in that code a request from the front end can't trip backend validation (famous last words etc). 

That said, I didn't test removing some frontend validation and seeing how the error displays (if at all).

I'm assuming without the changes the user would presumably get a 500 and a stack trace where this gives descriptive validation errors (displayed how though IDK will test tomorrow).

**Comment by dtuite** [comment]
Got it. All good to ship upstream as far as I'm concerned.
